nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -0,0 +1,330 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.mastodon;
+
+  env = {
+    RAILS_ENV = "production";
+    NODE_ENV = "production";
+
+    VAPID_PUBLIC_KEY = cfg.vapidPublicKey;
+    DB_USER = cfg.dbUser;
+    SMTP_LOGIN  = cfg.smtpLogin;
+
+    REDIS_HOST = cfg.redisHost;
+    REDIS_PORT = toString(cfg.redisPort);
+    DB_HOST = cfg.dbHost;
+    DB_PORT = toString(cfg.dbPort);
+    DB_NAME = cfg.dbName;
+    LOCAL_DOMAIN = cfg.localDomain;
+    SMTP_SERVER = cfg.smtpServer;
+    SMTP_PORT = toString(cfg.smtpPort);
+    SMTP_FROM_ADDRESS = cfg.smtpFromAddress;
+    PAPERCLIP_ROOT_PATH = "/var/lib/mastodon/public-system";
+    PAPERCLIP_ROOT_URL = "/system";
+    ES_ENABLED = if (cfg.elasticsearchHost != null) then "true" else "false";
+    ES_HOST = cfg.elasticsearchHost;
+    ES_PORT = toString(cfg.elasticsearchPort);
+  };
+
+in {
+
+  options = {
+    services.mastodon = {
+      enable = lib.mkEnableOption "Mastodon federated social network";
+      configureNginx = lib.mkOption {
+        description = "Use nginx as reverse proxy for mastodon";
```

Comment:
It would be great to have some documentation here of what the user has to do if they want to disable this and use a different reverse-proxy (such as Caddy).

Reply:
It's actually disabled by default.

Reply:
It's actually not so easy to tell if it should be enabled or disabled by default. If it stays enabled by default, I think we have to add a check that mastodon itself is enabled before setting the options.
I thought I had left it disabled by default, but now that I think about it it should probably stay enabled, because the service is not really usable without a correctly configured reverse proxy.

Reply:
I would be pretty surprised if I went to enable mastodon and ended up with nginx running without warning, given that I already have Caddy set up to listen on ports 80/443 and nginx would try to use those ports as well, which wouldn't work. I'd prefer to have this default to `false` with appropriate documentation.

It's also worth pointing out that the nginx configuration used here relies on the structured vhosts, and if I have nginx configured using a custom config file instead then it wouldn't work.

Mastodon may only document how to use nginx, but the actual proxy requirements can be inferred from the nginx config file and replicated using other reverse proxies. Even if you don't document the exact requirements here, you could link to the Mastodon documentation on how to set up the reverse proxy.

Reply:
As you wish, let's see if anyone else complains about it being disabled by default later. 
I'm just wondering where I should add that documentation.